_posts/2019-6-5-towards-reproducible-research-with-pytorch-hub.md

```diff
@@ -0,0 +1,164 @@
+
+```python
+model = torch.hub.load(..., force_reload=True)
+```
+
+We believe this will help to alleviate the burden of repetitive package releases by repo owners and instead allow them to focus more on their research. It also ensures that, as a user, you are getting the freshest available models.
+
+
+### Explore a loaded model
+
+Once you have a model from PyTorch Hub loaded, you can use the following workflow to find out the available methods that are supported as well as understand better what arguments are requires to run it.
+
+
+* ```dir(model)``` to see all available methods of the model. Let's take a look at bertForMaskedLM's available methods.
+
+```python
+>>> dir(model)
+>>>
+['forward'
+...
+'to'
+'state_dict',
+]
+```
+
+* ```help(model.forward)``` provides a view into what arguments are required to make your loaded model run
+
+```python
+>>> help(model.forward)
+>>>
+Help on method forward in module pytorch_pretrained_bert.modeling:
+forward(input_ids, token_type_ids=None, attention_mask=None, masked_lm_labels=None)
+...
+```
+
+Note that the ```*args```, ```**kwargs``` passed to hub.load() are used to *instantiate* a model.
+
+
+You can find a full script for predicting a masked word using the Bert model from PyTorch Hub [here](https://github.com/pytorch/hub/blob/master/huggingface_pytorch-pretrained-bert_bert.md).
+
+## Resources to get started
+
+* PyTorch Hub API documentation can be found [here](https://pytorch.org/docs/stable/hub.html).
+* Submit a model [here](https://github.com/pytorch/hub) for publication in PyTorch Hub.
+* Go to [https://pytorch.org/hub](https://pytorch.org/hub) to learn more about the available models.
+
+
+A BIG thanks to the folks at HuggingFace, fast.ai and Nvidia as well as Morgane Riviere (FAIR Paris) and lots of others for the help to bootstrap this effort!!
+
+
+## FAQ:
+
+**Q: If we would like to contribute a model that is already in the Hub but perhaps mine has better accuracy, should I still contribute?**
+A: Yes!! A next step for Hub is to implement an upvote/downvote system to surface the best models.
+
+**Q: Who hosts the model weights for PyTorch Hub?**
+A: You, as the contributor, are responsible to host the model weights. You can host your model in your favorite cloud storage or, if it fits within the limits, on GitHub.
+
+**Q: What if my model is trained on private data? Should I still contribute this model?**
+A: No! PyTorch Hub is centered around open research and that extends to the usage of open datasets to train these models on. If a pull request for a proprietary model is submitted, we will kindly ask that you resubmit a model trained on something open and available.
+
+**Q: Where are my downloaded models saved?**
+A: The locations are used in the order of:
+
+* Calling hub.set_dir(<PATH_TO_HUB_DIR>)
```

Comment:
* Calling `hub.set_dir()`
* `$TORCH_HOME/hub`, if environment variable `$TORCH_HOME` is set.
* `$XDG_CACHE_HOME/torch/hub`, if environment variable `$XDG_CACHE_HOME` is set.
* `~/.cache/torch/hub`